### PR TITLE
Use throw_exception overload taking source_location

### PR DESCRIPTION
**Pull request by pdimov**

### CODE REVIEW

**Comment by Lastique** [comment]
Can we reverse the order of exception and source location in `boost::throw_exception`?

The reason is I would like to have an overload like this:

```
template< typename E, typename... Args >
void throw_exception(source_location const& loc, Args&&... args);
```

which would encapsulate the construction of `E` from `args`. I would also mark it as cold and noinline. This would reduce the need for custom overloads like the ones in Boost.Sync.

**Comment by pdimov** [comment]
This made sense to me, but on reflection...

The implementation would be something like
```
template< typename E, typename... Args >
void throw_exception(source_location const& loc, Args&&... args)
{
    boost::throw_exception(loc, E(std::forward<Args>(args)...));
}
```
but this overload is a better match than the other one, so we'll have to SFINAE it out. In addition, I don't like the ambiguities that it introduces in the case `throw_exception(loc)` - in general it's better for the first argument to always have the same meaning across all overloads.

Not insisting on naming this overload `throw_exception` avoids all these problems. We can just call it `throw_exception_` for instance.

**Comment by Lastique** [comment]
I'm not sure I understand. The full set of overloads would be:

```
// Legacy one
template< typename E >
void throw_exception(E const& e);

// The one you added, with reversed arguments
template< typename E >
void throw_exception(boost::source_location const& loc, E const& e);

// The one I proposed
template< typename E, typename... Args >
void throw_exception(boost::source_location const& loc, Args&&... args);
```

Note that the last one is only callable when you explicitly specify template parameter `E`. The other two are unambiguous.

Or do you mean the first and the last overloads? We should require `E` to derive from `std::exception`, we already do that in via `throw_exception_assert_compatibility` in the hard way. `throw_exception(loc)` won't compile, but something like `throw_exception< std::bad_alloc >(loc)` should and will.

Another name would work, if it must be, though I would prefer to avoid unnecessary mangling, like a trailing underscore.

**Comment by pdimov** [comment]
Yes, going over all the possibilities, it does look like everything would work, but I still find the combination too fragile for my taste.

**Comment by pdimov** [comment]
Will we have additional overloads in C++03 mode?

**Comment by Lastique** [comment]
Well, Boost.Sync has these overloads in C++03. I think, up to 3 arguments would be fine, to be able to throw `system_error`.

**Comment by pdimov** [comment]
This introduces a bit of a problem; if we include the one-argument overload
```
template< typename E, class A1 >
void throw_exception(boost::source_location const& loc, A1 const& a1)
{
    throw_exception(loc, E(a1));
}
```
then `throw_exception<E>(loc, e)` becomes ambiguous. And if we don't, `throw_exception<E>(loc, x)` doesn't work when `E(x)` is explicit.

The latter might be an acceptable compromise.

**Comment by pdimov** [comment]
Why is `throw_exception<E>(loc, args...)` superior to `throw_exception(E(args...), loc)`? Seems to me they are basically identical in number of tokens.

**Comment by Lastique** [comment]
> `throw_exception<E>(loc, e)` becomes ambiguous.

It shouldn't be if `e` has a type other than `E`. Granted, that would be more difficult to test in the variadic version.

> Why is `throw_exception<E>(loc, args...)` superior to `throw_exception(E(args...), loc)`?

It is better because it moves exception creation out of the calling code, which may be a hot path. Exception creation may be large in code size, so it is desirable to move it away from the hot code and possibly move it to the cold code section. Ideally, I'd like to see `throw_exception` to be a few instructions of argument loading and a single call/jump.

Which makes me wonder, do we need column in `source_location`? Are there compilers that provide this information? Because if not, it'd be probably better to remove it to avoid the overhead. Also, it may be better to pass `source_location` by value so that it is passed in registers.

**Comment by pdimov** [comment]
No, I don't think there are any compilers that provide a column. I've just copied std::source_location. Since it's already two pointers and an uint32_t, it doesn't fit in registers either way, so there's no real benefit to deviate, although I suppose I could just have hardcoded column() to `return 0`.

**Comment by Lastique** [comment]
I wasn't aware it was a counterpart of a standard component. In that case keeping the standard public interface makes sense.

Though I have a hard time imagining what the column would be if `source_location` is constructed as part of `BOOST_CURRENT_LOCATION`. I mean, if there is a `__COLUMN__` macro, what would it's value be after `BOOST_CURRENT_LOCATION` expansion. I'm also not sure how useful it would be. Looks like overengineering on the standard part. Maybe the `return 0` approach would be better until we have at least one compiler that supports this feature.

**Comment by pdimov** [comment]
The standard component (it's in C++20) has a magic `static source_location current();` function that returns the current source location. No idea what the column is supposed to be there. The intended use is
```
void f( int x, std::source_location const& loc = std::source_location::current() );
```
so when the user calls `f(5)`, the current location is magically inserted, without need for a macro. So in a hypothetical C++20 Boost, throw_exception could have been
```
template<class E> void throw_exception(E&& e, source_location const& loc = source_location::current());
```

I haven't tried to emulate `source_location::current` even for the case (recent gcc) where it's possible to achieve some approximation of it by using `__builtin_FILE` etc.

I agree that the column seems like over-engineering.

**Comment by Lastique** [comment]
So, what's your decision on the order of arguments of `throw_exception`? Should I wait merging this PR?

**Comment by pdimov** [comment]
I think I'll retain the current order and not introduce variadic overloads under the same name.

File: include/boost/sync/detail/throw_exception.hpp [hunk 1]
```diff
@@ -18,9 +18,6 @@
 #include <boost/throw_exception.hpp>
 #include <boost/preprocessor/seq/enum.hpp>
 #include <boost/sync/detail/config.hpp>
-#if defined(BOOST_EXCEPTION_DISABLE)
-#include <boost/current_function.hpp>
```
Comment (Lastique) [inline include/boost/sync/detail/throw_exception.hpp:22]: `source_location.hpp` include is missing.

Comment (pdimov) [reply include/boost/sync/detail/throw_exception.hpp:22]: Added.

**Comment by Lastique** [comment]
Thanks.

**Pull request merged**